README.md
<!-- README.md is generated from README.Rmd. Please edit that file, then:
rmarkdown::render("README.Rmd")
-->

# PRISMAstatement

[![CRAN
status](https://www.r-pkg.org/badges/version/PRISMAstatement)](https://cran.r-project.org/package=PRISMAstatement)
[![Travis build
status](https://travis-ci.org/jackwasey/PRISMAstatement.svg?branch=master)](https://travis-ci.org/jackwasey/PRISMAstatement)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/jackwasey/PRISMAstatement?branch=master&svg=true)](https://ci.appveyor.com/project/jackwasey/PRISMAstatement)
[![Coverage
status](https://codecov.io/gh/jackwasey/PRISMAstatement/branch/master/graph/badge.svg)](https://codecov.io/github/jackwasey/PRISMAstatement?branch=master)
[![lifecycle](https://img.shields.io/badge/lifecycle-stable-blue.svg)](https://www.tidyverse.org/lifecycle/#stable)

PRISMAstatement is a simple package which does one thing: generate
publication quality flow charts of study inclusions and exclusions
following recommendations of the PRISMA statement. See the vignettes
(`vignette("PRISMA")`, `vignette("exclusionflowcharts")`) for more
details.

## Installation

You can install the released version of PRISMAstatement from
[CRAN](https://CRAN.R-project.org/package=PRISMAstatement) with:

``` r
install.packages("PRISMAstatement")
```

## Example

``` r
library(PRISMAstatement)
prisma(found = 750,
       found_other = 123,
       no_dupes = 776, 
       screened = 776, 
       screen_exclusions = 13, 
       full_text = 763,
       full_text_exclusions = 17, 
       qualitative = 746, 
       quantitative = 319)
```

<!--html_preserve-->

<div id="htmlwidget-e9627b9d67ff96545c30" class="grViz html-widget" style="width:100%;height:480px;">

</div>

<script type="application/json" data-for="htmlwidget-e9627b9d67ff96545c30">{"x":{"diagram":"digraph prisma {\n    node [shape=\"box\", fontsize = 10];\n    graph [splines=ortho, nodesep=1, dpi = 72]\n    a -> nodups;\n    b -> nodups;\n    a [label=\"Records identified through\ndatabase searching\n(n = 750)\"];\n    b [label=\"Additional records identified\nthrough other sources\n(n = 123)\"]\n    nodups -> incex;\n    nodups [label=\"Records after duplicates removed\n(n = 776)\"];\n    incex -> {ex; ft}\n    incex [label=\"Records screened\n(n = 776)\"];\n    ex [label=\"Records excluded\n(n = 13)\"];\n    {rank=same; incex ex}\n    ft -> {qual; ftex};\n    ft [label=\"Full-text articles assessed\nfor eligibility\n(n = 763)\"];\n    {rank=same; ft ftex}\n    ftex [label=\"Full-text articles excluded,\nwith reasons\n(n = 17)\"];\n    qual -> quant\n    qual [label=\"Studies included in qualitative synthesis\n(n = 746)\"];\n    quant [label=\"Studies included in\nquantitative synthesis\n(meta-analysis)\n(n = 319)\"];\n  }","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>

<!--/html_preserve-->


NEWS.md
# PRISMAstatement news

## Version 1.1.0

* Function to make generic flow charts with exclusions

## Version 1.0.5

* Added JOSS draft paper
* fixed various build ignore and vignette issues
* updated vignette with content from JOSS draft

## Version 1.0.4

* Fix minor bug in display of text in one of the PRISMA boxes
* Added function to give an single PDF with the diagram for publication.

## Version 1.0.3

* added option of an additional box describing the (calculated) number of
  duplicates removed.
* added option to customize the labels in each box
* added numerical constraints on the inputs with errors and warnings. See
  the vignette and `?prisma` for details.

## Version 1.0.2

* added DPI option and demonstration of its danger in vignette
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(PRISMAstatement)

test_check("PRISMAstatement")


tests/testthat/test-prisma.r
context("basic")

test_that("prisma runs without error", {
  expect_error(
    prisma(found = 750,
           found_other = 123,
           no_dupes = 776,
           screened = 776,
           screen_exclusions = 13,
           full_text = 763,
           full_text_exclusions = 17,
           qualitative = 746,
           quantitative = 319),
    regexp = NA)
})

test_that("prisma passes values to grViz", {
  expect_error(
    prisma(found = 750,
           found_other = 123,
           no_dupes = 776,
           screened = 776,
           screen_exclusions = 13,
           full_text = 763,
           full_text_exclusions = 17,
           qualitative = 746,
           quantitative = 319,
           width = 800, height = 800),
    regexp = NA)
})

test_that("all values could be zero", {
  expect_error(
    prisma(found = 0,
           found_other = 0,
           no_dupes = 0,
           screened = 0,
           screen_exclusions = 0,
           full_text = 0,
           full_text_exclusions = 0,
           qualitative = 0,
           quantitative = 0,
           width = 800, height = 800),
    regexp = NA)
})

test_that("can vary DPI", {
  expect_error(
    prisma(found = 0,
           found_other = 0,
           no_dupes = 0,
           screened = 0,
           screen_exclusions = 0,
           full_text = 0,
           full_text_exclusions = 0,
           qualitative = 0,
           quantitative = 0,
           width = 800, height = 800,
           dpi = 300),
    regexp = NA)
})

test_that("dupes box", {
  expect_error(
    prisma(found = 0,
           found_other = 0,
           no_dupes = 0,
           screened = 0,
           screen_exclusions = 0,
           full_text = 0,
           full_text_exclusions = 0,
           qualitative = 0,
           quantitative = 0,
           extra_dupes_box = TRUE),
    regexp = NA)
})

test_that("change a label", {
  expect_error(
    prisma(found = 0,
           found_other = 0,
           no_dupes = 0,
           screened = 0,
           screen_exclusions = 0,
           full_text = 0,
           full_text_exclusions = 0,
           qualitative = 0,
           quantitative = 0,
           labels = list(screened = "SCREENED")),
    regexp = NA)
})

test_that("numbers must work together", {
  expect_error(prisma(1, 2, 3, 4, 5, 6, 7, 8, 9))
  expect_warning(prisma(1000, 20, 270, 270, 10, 260, 19, 240, 107))
  expect_warning(prisma(1000, 20, 270, 270, 269, 260, 20, 240, 107))
})

context("pdf")

test_that("create PDF output", {
  if (!requireNamespace("DiagrammeRsvg", quietly = TRUE) ||
      !requireNamespace("rsvg", quietly = TRUE))
    skip("need DiagrammeRsvg and rsvg for this test")
  tmp_pdf <- tempfile()
  on.exit(unlink(tmp_pdf), add = TRUE)
  prsm <- prisma(1000, 20, 270, 270, 10, 260, 20, 240, 107)
  PRISMAstatement:::prisma_pdf(prsm, tmp_pdf)
  expect_true(file.size(tmp_pdf) > 0)
})


tests/testthat/test-digraph.r
context("test generic digraph flow")

test_that("numbers only digraph", {
  dot <- flow_exclusions_dot(c(1000, 300, 150, 75, 38))
  testthat::expect_error(DiagrammeR::grViz(dot), regexp = NA)
})

test_that("labels and numbers digraph", {
  dot <-
    flow_exclusions_dot(
      incl_counts = c(972, 132, 77, 14),
      total_label = "Total Screened",
      incl_labels = c("Consented", "Completed Study", "BMI <= 30"),
      excl_labels = c("Declined Consent", "Failed to Complete", "BMI > 30")
    )
  testthat::expect_error(DiagrammeR::grViz(dot), regexp = NA)
})

test_that("percents", {
  dot <- flow_exclusions_dot(incl_counts = c(1000, 300, 150),
                            percent_of_total = TRUE)
  testthat::expect_error(DiagrammeR::grViz(dot), regexp = NA)
  dot <- flow_exclusions_dot(incl_counts = c(1000, 300, 150),
                            percent_of_prev = TRUE)
  testthat::expect_error(DiagrammeR::grViz(dot), regexp = NA)
})


mergebib.sh
#!/bin/bash

#cp vignettes/icd.bib vignettes/all.bib
#cat vignettes/other.bib >> vignettes/all.bib
#cat vignettes/gplv3.bib >> vignettes/all.bib

command -v bibtool >/dev/null 2>&1 || { echo >&2 "I require bibtool but it's not installed. Aborting."; exit 1; }

if [ -d vignettes ]; then
  pushd vignettes
  trap popd EXIT
fi

bibtool -s -r biblatex r.bib PRISMA.bib -- preserve.key.case=on > all.bib

# make sure there are no unescaped underscores in the URLs:
if grep '[^\]_' all.bib | grep -v ^@
then
  echo "Underscores found in all.bib entries"

  echo "attempting to fix"
  perl -pi -e 's/(url.*[^\\])_(.*)/\1\\_\2/g' all.bib
  perl -pi -e 's/(url.*[^\\])_(.*)/\1\\_\2/g' all.bib
  perl -pi -e 's/(url.*[^\\])_(.*)/\1\\_\2/g' all.bib
  perl -pi -e 's/(url.*[^\\])_(.*)/\1\\_\2/g' all.bib
  perl -pi -e 's/(url.*[^\\])_(.*)/\1\\_\2/g' all.bib
  perl -pi -e 's/(url.*[^\\])_(.*)/\1\\_\2/g' all.bib
  perl -pi -e 's/(url.*[^\\])_(.*)/\1\\_\2/g' all.bib
  perl -pi -e 's/(url.*[^\\])_(.*)/\1\\_\2/g' all.bib
  exit 1
fi




inst/JOSS/paper.sh
#!/bin/bash

if false
then
echo "Doing full pandoc for JOSS"
pandoc \
  --verbose \
  -V repository="repo" \
  -V archive_doi="doi" \
  -V paper_url="purl" \
  -V formatted_doi="fdoi" \
  -V review_issue_url="revisurl" \
  -V graphics="true" \
  -V issue="issue" \
  -V volume="volume" \
  -V page="page" \
  -V joss_logo_path="/usr/local/src/git/whedon/resources/joss-logo.png" \
  -V year="year" \
  -V submitted="sub" \
  -V published="pub" \
  -V citation_author="jackwasey" \
  -V paper_title="papertitle" \
  -V geometry:margin=1in \
  -o paper.pdf \
  --pdf-engine=xelatex \
  --filter pandoc-citeproc paper.md \
  --template "/usr/local/src/git/whedon/resources/latex.template"
else
  echo "Doing test pandoc for JOSS"
  pandoc \
    --verbose \
    -V geometry:margin=1in \
    -V joss_logo_path="/usr/local/src/git/whedon/resources/joss-logo.png" \
    -o paper.pdf \
    --pdf-engine=xelatex \
    --filter pandoc-citeproc paper.md \
    --template "/usr/local/src/git/whedon/resources/latex.template"
fi



R/flow-exclusions.r
#' Generic exclusions flow chart (beta)
#'
#' Generate the DOT description for a series of inclusions and
#' exclusions
#' @examples
#' flow_exclusions(c(1000, 300, 150, 75, 38))
#' dot <-
#'   flow_exclusions_dot(
#'     incl_counts = c(972, 132, 77, 14),
#'     total_label = "Total Screened",
#'     incl_labels = c("Consented", "Completed Study", "BMI <= 30"),
#'     excl_labels = c("Declined Consent", "Failed to Complete", "BMI > 30")
#'     )
#' DiagrammeR::grViz(dot)
#' flow_exclusions(c(1000, 300, 150, 75, 38),
#'                 percent_of_total = TRUE,
#'                 show_count = FALSE)
#' flow_exclusions(c(100000, 3000, 1666, 411, 38),
#'                 percent_of_prev = TRUE,
#'                 percent_args = list(decimal.mark = ","),
#'                 format_args = list(scientific = TRUE),
#'                 show_count = TRUE)
#' @param percent_args `list` of arguments to pass to `scales::percent` for
#'   formatting percentages. Default is an empty list.
#' @param format_args `list` of arguments to pass to `base::format` for
#'   formatting counts. Default is an empty list.
#' @param ... arguments passed to `flow_exclusions_dot`
#' @export
#' @md
#' @keywords internal
flow_exclusions <- function(
  incl_counts,
  total_label = "Total",
  incl_labels = LETTERS[seq_len(length(incl_counts) - 1)],
  excl_labels = paste("Not ", incl_labels),
  show_count = TRUE,
  percent_of_total = FALSE,
  percent_of_prev = FALSE,
  percent_args = list(),
  format_args = list(),
  font_size = NULL, dpi = NULL,
  ...)
DiagrammeR::grViz(
  flow_exclusions_dot(
    incl_counts = incl_counts,
    total_label = total_label,
    incl_labels = incl_labels,
    excl_labels = excl_labels,
    show_count = show_count,
    percent_of_total = percent_of_total,
    percent_of_prev = percent_of_prev,
    percent_args = percent_args,
    format_args = format_args,
    font_size = font_size, dpi = dpi, ...)
)

#' @describeIn flow_exclusions Just generate the DOT text
#' @export
#' @keywords internal
flow_exclusions_dot <- function(
  incl_counts,
  total_label = "Total",
  incl_labels = LETTERS[seq_len(length(incl_counts) - 1)],
  excl_labels = paste("Not ", incl_labels),
  show_count = TRUE,
  percent_of_total = FALSE,
  percent_of_prev = FALSE,
  percent_args = list(),
  format_args = list(),
  font_size = NULL, dpi = NULL,
  ...) {
  is.wholenumber <-
    function(x) abs(x - round(x)) < .Machine$double.eps
  if (any(!is.wholenumber(incl_counts))) {
    warning("Some counts are not whole numbers, rounding them")
    incl_counts <- round(incl_counts)
  }
  excl_counts <- incl_counts[-length(incl_counts)] - incl_counts[-1]
  total_count <- incl_counts[1]
  stopifnot(length(total_label) == 1)
  stopifnot(length(incl_counts) == length(incl_labels) + 1,
            length(incl_counts) == length(excl_labels) + 1)
  stopifnot((is.numeric(font_size) && length(font_size) == 1) ||
              is.null(font_size))
  stopifnot((is.numeric(dpi) && length(dpi) == 1) ||
              is.null(dpi))
  stopifnot(!(percent_of_total && percent_of_prev))

  fixchar <- function(y) gsub(pattern = "[[:punct:][:space:]]",
                              replacement = "", x = y, perl = TRUE)
  total_nm <- fixchar(total_label)
  incl_nm <- fixchar(incl_labels)
  excl_nm <- fixchar(excl_labels)
  # to avoid duplicates being created accidentlly, add unique id to each
  incl_nm <- paste0(incl_nm, "_", LETTERS[seq_along(incl_nm)])
  excl_nm <- paste0(excl_nm, "_", letters[seq_along(excl_nm)])
  stopifnot(!anyDuplicated(c(total_nm, incl_nm, excl_nm)))
  if (any(c(show_count, percent_of_total, percent_of_prev))) {
    total_label <- paste0(total_label, "\n")
    incl_labels <- paste0(incl_labels, "\n")
    excl_labels <- paste0(excl_labels, "\n")
    if (show_count) {
      total_label <- paste0(total_label,
                            do.call(format, c(list(total_count), format_args)))
      incl_labels <- paste0(incl_labels,
                            do.call(format, c(list(incl_counts[-1]), format_args)))
      excl_labels <- paste0(excl_labels,
                            do.call(format, c(list(excl_counts), format_args)))
    }
    if (percent_of_total || percent_of_prev) {
      if (percent_of_total) {
        incl_frac <- incl_counts[-1] / total_count
        excl_frac <- excl_counts / total_count
      } else {
        incl_frac <- incl_counts[-1] / incl_counts[-length(incl_counts)]
        excl_frac <- excl_counts / incl_counts[-length(incl_counts)]
      }
      hndr_perc <- do.call(scales::percent, c(list(1), percent_args))
      incl_perc <- do.call(scales::percent, c(list(incl_frac), percent_args))
      excl_perc <- do.call(scales::percent, c(list(excl_frac), percent_args))
      gap <- ifelse(show_count, " ", "")
      total_label <- paste0(total_label, gap, "(", hndr_perc, ")")
      incl_labels <- paste0(incl_labels, gap, "(", incl_perc, ")")
      excl_labels <- paste0(excl_labels, gap, "(", excl_perc, ")")
    }
    # end any percentage
  }
  # end any numbers at all
  head_template <- '
    digraph prisma {
    node [shape="box" %s];
    graph [splines=ortho, nodesep=1 %s]
    '
  dot_head <- sprintf(
    head_template,
    ifelse(is.null(font_size), "", sprintf(", fontsize = %d", font_size)),
    ifelse(is.null(dpi), "", sprintf(", dpi = %d", dpi))
  )
  label_template <- "%s [label=\"%s\"];"
  arrow_template <- "%s -> {%s; %s};"
  rank_template <- "{rank=same; %s %s}"
  dot_labels <- paste(sprintf(label_template,
                              c(total_nm, incl_nm, excl_nm),
                              c(total_label, incl_labels, excl_labels)),
                      collapse = "\n")
  dot_arrows <- paste(sprintf(arrow_template,
                              c(total_nm, incl_nm[-length(incl_nm)]),
                              incl_nm,
                              excl_nm),
                      collapse = "\n")
  dot_rank <- paste(sprintf(rank_template,
                            c(total_nm, incl_nm[-length(incl_nm)]),
                            excl_nm),
                    collapse = "\n")
  paste(dot_head, dot_labels, dot_arrows, dot_rank, "}", sep = "\n")
}


R/prisma.r
#' Generate a PRISMA statement flow chart
#'
#' Generate PRISMA statement flow chart for use in retrospective medical
#' research. Almost all arguments are mandatory, as they are in the recommended
#' PRISMA statement.
#' @param found Records found through database searching
#' @param found_other Additional records identified through other sources
#' @param no_dupes Records after duplicates removed
#' @param screened Records screened
#' @param screen_exclusions Records excluded
#' @param full_text Full-text articles assessed for eligibility
#' @param full_text_exclusions Full-text articles excluded with reasons
#' @param qualitative Studies included in qualitative analysis
#' @param quantitative Studies included in quantitative synthesis
#'   (meta-analysis)
#' @param labels \code{NULL} is the default, but if a named list of character
#'   strings, the box matching each name will get the corresponding label. See
#'   examples.
#' @param extra_dupes_box Single logical value, default is \code{FALSE} which
#'   corresponds to the example 2009 PRISMA Statement Flow Chart. If
#'   \code{TRUE}, then an additional box will be presented indicating the number
#'   of duplicates removed, calculated from the other numbers.
#' @param ... Further arguments are passed to \code{grViz}
#' @param dpi Dots per inch, 72 is the default here, and in \code{DiagrammeR}
#'   itself it claims to be 96. Varying the DPI (which is done in the DOT file)
#'   unfortunately does not get detected by the downstream processing by the
#'   'htmlwidgets' package. To overcome this, the user can add `height` and
#'   `width` arguments which are passed through. It is easy to for scaled graphs
#'   to fall off the canvas, or be crushed into the top-left corner, and
#'   unfortunately this requires trial and error. Increasing DPI over 72 with
#'   this setting tends to truncate the graph. On the other hand, leaving the
#'   DPI at 72 and increasing both height and width appears to consistently give
#'   higher resolution images.
#' @param font_size integer font size in points, default is 10. `DiagrammeR` via
#'   `htmlwidgets` should scale the boxes to include the text no matter what
#'   size font is used. However, the heuristics are not perfect, so tweaking the
#'   font size here may help prepare for publication.
#' @source \url{http://prisma-statement.org/PRISMAStatement/FlowDiagram}
#' @examples
#' prisma(1000, 20, 270, 270, 10, 260, 20, 240, 107)
#' prisma(1000, 20, 270, 270, 10, 260, 20, 240, 107,
#'        labels = list(found = "FOUND"))
#' prisma(1000, 20, 270, 270, 10, 260, 20, 240, 107, dpi = 24)
#' prisma(1000, 20, 270, 270, 10, 260, 20, 240, 107, extra_dupes_box = TRUE)
#' # vary the font size
#' prisma(1000, 20, 270, 270, 10, 260, 20, 240, 107, font_size = 6)
#' prisma(1000, 20, 270, 270, 10, 260, 20, 240, 107, font_size = 60)
#' # giving impossible numbers should cause an error
#' \dontrun{
#'   prisma(1, 2, 3, 4, 5, 6, 7, 8, 9)
#' # giving unlikely numbers should cause a warning
#'   prisma(1000, 20, 270, 270, 10, 260, 19, 240, 107)
#'   prisma(1000, 20, 270, 270, 269, 260, 20, 240, 107)
#' }
#' @md
#' @export
prisma <- function(found,
                   found_other,
                   no_dupes,
                   screened,
                   screen_exclusions,
                   full_text,
                   full_text_exclusions,
                   qualitative,
                   quantitative = NULL,
                   labels = NULL,
                   extra_dupes_box = FALSE,
                   ...,
                   dpi = 72,
                   font_size = 10) {
  DiagrammeR::grViz(
    prisma_graph(found = found,
                 found_other = found_other,
                 no_dupes = no_dupes,
                 screened = screened,
                 screen_exclusions = screen_exclusions,
                 full_text = full_text,
                 full_text_exclusions = full_text_exclusions,
                 qualitative = qualitative,
                 quantitative = quantitative,
                 labels = labels,
                 extra_dupes_box = extra_dupes_box,
                 dpi = dpi,
                 font_size = font_size,
                 ...)
  )
}

#' @describeIn prisma Generate the `dot` graph text
#' @export
prisma_graph <- function(found,
                         found_other,
                         no_dupes,
                         screened,
                         screen_exclusions,
                         full_text,
                         full_text_exclusions,
                         qualitative,
                         quantitative = NULL,
                         labels = NULL,
                         extra_dupes_box = FALSE,
                         ...,
                         dpi = 72,
                         font_size = 10) {
  stopifnot(length(found) == 1)
  stopifnot(length(found_other) == 1)
  stopifnot(length(no_dupes) == 1)
  stopifnot(length(screened) == 1)
  stopifnot(length(screen_exclusions) == 1)
  stopifnot(length(full_text) == 1)
  stopifnot(length(full_text_exclusions) == 1)
  stopifnot(length(qualitative) == 1)
  stopifnot(is.null(quantitative) || length(quantitative) == 1)
  # each number should be a non-negative integer (but may be 'numeric' type)
  stopifnot(found == floor(found))
  stopifnot(found_other == floor(found_other))
  stopifnot(no_dupes == floor(no_dupes))
  stopifnot(screened == floor(screened))
  stopifnot(screen_exclusions == floor(screen_exclusions))
  stopifnot(full_text == floor(full_text))
  stopifnot(full_text_exclusions == floor(full_text_exclusions))
  stopifnot(qualitative == floor(qualitative))
  stopifnot(is.null(quantitative) || quantitative == floor(quantitative))
  stopifnot(found >= 0)
  stopifnot(found_other >= 0)
  stopifnot(no_dupes >= 0)
  stopifnot(screened >= 0)
  stopifnot(screen_exclusions >= 0)
  stopifnot(full_text >= 0)
  stopifnot(full_text_exclusions >= 0)
  stopifnot(qualitative >= 0)
  stopifnot(is.null(quantitative) || quantitative >= 0)
  # can't have more articles at any stage
  stopifnot(no_dupes <= found + found_other)
  stopifnot(screened <= no_dupes)
  stopifnot(full_text <= screened)
  stopifnot(qualitative <= full_text)
  stopifnot(quantitative <= qualitative)
  # exclusions can't be greater than what they excluded from
  stopifnot(screen_exclusions <= screened)
  stopifnot(full_text_exclusions <= full_text)
  if (screened - screen_exclusions != full_text)
    warning("After screening exclusions, a different number of remaining ",
            "full-text articles is stated.")
  if (full_text - full_text_exclusions != qualitative)
    warning("After full-text exclusions, a different number of remaining ",
            "articles for qualitative synthesis is stated.")
  dupes <- found + found_other - no_dupes
  labels_orig <- list(
    found = pnl("Records identified through",
                "database searching",
                paren(found)),
    found_other = pnl("Additional records identified",
                      "through other sources",
                      paren(found_other)),
    no_dupes = pnl("Records after duplicates removed", paren(no_dupes)),
    dupes = pnl("Duplicates excluded", paren(dupes)),
    screened = pnl("Records screened", paren(screened)),
    screen_exclusions = pnl("Records excluded", paren(screen_exclusions)),
    full_text = pnl("Full-text articles assessed",
                    "for eligibility",
                    paren(full_text)),
    full_text_exclusions =
      pnl("Full-text articles excluded,",
          "with reasons",
          paren(full_text_exclusions)),
    qualitative = pnl("Studies included in qualitative synthesis",
                      paren(qualitative)),
    quantitative = pnl("Studies included in",
                       "quantitative synthesis",
                       "(meta-analysis)",
                       paren(quantitative))
  )
  for (l in names(labels))
    labels_orig[[l]] <- labels[[l]]
  labels <- labels_orig
  dupes_box <- sprintf(
    'nodups -> incex;
    nodups [label="%s"];',
    labels$no_dupes)
  if (extra_dupes_box)
    dupes_box <- sprintf(
      'nodups -> {incex; dups};
       nodups [label="%s"];
       dups [label="%s"]; {rank=same; nodups dups}',
      labels$no_dupes, labels$dupes)

  dot_template <- 'digraph prisma {
    node [shape="box", fontsize = %d];
    graph [splines=ortho, nodesep=1, dpi = %d]
    a -> nodups;
    b -> nodups;
    a [label="%s"];
    b [label="%s"]
    %s
    incex -> {ex; ft}
    incex [label="%s"];
    ex [label="%s"];
    {rank=same; incex ex}
    ft -> {qual; ftex};
    ft [label="%s"];
    {rank=same; ft ftex}
    ftex [label="%s"];
    qual -> quant
    qual [label="%s"];
    quant [label="%s"];
  }'
    sprintf(dot_template,
            font_size,
            dpi,
            labels$found,
            labels$found_other,
            dupes_box,
            labels$screened,
            labels$screen_exclusions,
            labels$full_text,
            labels$full_text_exclusions,
            labels$qualitative,
            labels$quantitative)
}

paren <- function(n)
  sprintf("(n = %d)", n)

pnl <- function(...)
  paste(..., sep = "\n")

#' Make PDF of the plot
#'
#' This makes a PDF file which can be included by knitr Sweave.
#' @param x output of call to \code{prisma}
#' @param filename path of output file
#' @importFrom utils capture.output
#' @examples
#' \dontrun{
#' g <- prisma(9, 8, 7, 6, 5, 4, 3, 2, 1)
#' prisma_pdf(g, "test.pdf")
#' knitr::include_graphics("test.pdf")
#' }
#' @keywords internal
prisma_pdf <- function(x, filename = "prisma.pdf") {
  if (!requireNamespace("DiagrammeRsvg", quietly = TRUE) ||
      !requireNamespace("rsvg", quietly = TRUE)) {
    stop("DiagrammeRsvg and rsvg are both required for this prisma_pdf")
  }
  utils::capture.output({
    rsvg::rsvg_pdf(svg = charToRaw(DiagrammeRsvg::export_svg(x)),
                   file = filename)
  })
  invisible()
}

#' @describeIn prisma_pdf Export using any conversion function offered by `rsvg`
#' @param rsvg_fun Function from `rsvg` default being `rsvg::rsvg_png`
#' @param ... Passed to `rsvg_fun`
#' @examples
#' \dontrun{
#' g_dot <- prisma_graph(9, 8, 7, 6, 1, 5, 1, 4, 1)
#' prisma_export(g_dot, "test.png", rsvg_fun = rsvg::rsvg_png)
#' }
#' @export
prisma_export <- function(x,
                          filename = "prisma.png",
                          rsvg_fun = rsvg::rsvg_png,
                          ...) {
  if (!requireNamespace("DiagrammeRsvg", quietly = TRUE) ||
      !requireNamespace("rsvg", quietly = TRUE)) {
    stop("DiagrammeRsvg and rsvg are both required for exporting PRISMA charts")
  }
  DiagrammeR::export_graph(x, file_name = filename, )
  utils::capture.output({
    rsvg_fun(svg = charToRaw(DiagrammeRsvg::export_svg(x)),
             file = filename)
  })
  invisible()
}
